### PR TITLE
New Nose2 Fathead

### LINKED ISSUE #461: Python: Create New Nose Library Fathead
# Nose Fathead

Contribute to DuckDuckGo!

screen shot 2016-10-09 at 8 57 29 pm
### What you will need to know?

Either Python (Preferrable for this mission), Ruby, Perl or JavaScript for Fatheads. You will download the Python redis client source and create an **output.txt** file containing the parsed documentation. Read more about Fatheads here: 
https://docs.duckduckhack.com/resources/fathead-overview.html
### The source :rocket:

https://nose.readthedocs.io/en/latest/
### The goal

Help _improve Python language search results_ on DuckDuckGo as part of our Programming Mission! We're trying to reach 100% Instant Answer (IA) coverage for Python-related searches by creating new Instant Answers or improving existing ones.

Some relevent examples:
- [Django Built-ins](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/django_built_ins)
- [Redis](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/redis_commands)
- [Python documentation](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/python)

**Fathead Example (Django Built-ins)**

screen shot 2016-10-09 at 8 59 24 pm
### The steps
- [x] Claim this issue (Comment Below)
- [ ] Please review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] Set up codio environment (See: https://docs.duckduckhack.com/welcome/setup-dev-environment.html) and fork this repository
- [ ] Create the Fathead
- [ ] Create a PR
- [ ] Ping @pjhampton for a review
### Questions?
- [Documentation](https://docs.duckduckhack.com/)
- [The Forum](https://forum.duckduckhack.com/c/programming/python)
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) 
- Ping one of the [Community Leaders](https://duck.co/help/community/community-leaders) or @pjhampton

### PR DESCRIPTION
**Pull request by rnkaufman**
(This PR is currently a work in progress)

## Type of Change

- [x] New Instant Answer

## Description of new Instant Answer, or changes

This instant answer gives information about the Nose2 testing library for Python.

## Related Issues and Discussions

Fixes #461 

## People to notify

@tagawa 
@pjhampton 

---

Forum Topic: 
https://forum.duckduckhack.com/t/nose-docs-fathead/1034

IA Page: https://duck.co/ia/view/nose

### CODE REVIEW

**Comment by daxtheduck** [comment]
@rnkaufman Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by rnkaufman** [comment]
This is PR is currently a work in progress--I would like to clean up the code a bit and add support for more areas of the Nose2 documentation.

**Comment by moollaza** [comment]
@rnkaufman just checking in -- is this ready for a review? If not, please let us know when it is!

**Comment by rnkaufman** [comment]
Hi @moollaza if I can get some comments for now, that would be great! I just wanted to expand on it, but if there are things you see that you'd like to comment on now, I'd really appreciate it!

File: lib/fathead/nose2/parse.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+require 'open-uri'
+
+NOSE_HOME = "http://nose2.readthedocs.io/en/latest/getting_started.html"
+topics = ARGV
+
+topics.each do |topic|
+  url = "http://nose2.readthedocs.io/en/latest/getting_started.html##{topic}"
+  doc = Nokogiri::HTML(open(url))
+  abstract = doc.css("##{topic}").css("p").to_s
+  abstract = abstract.gsub("\n", " ").gsub(/class\=\S{1,2}[a-z]*(\s[a-z]*)?\S{1,2}/, "")
```
Comment (moollaza) [inline lib/fathead/nose2/parse.rb]: @rnkaufman You'll need to keep `<span>` and `<a>` tags out of the abstract. 

Code snippets should be wrapped in `<pre><code>`. Text content should be wrapped in `<p>`

The whole abstract needs to be wrapped in `<section class="prog__container"></section>`

Checkout the Fathead Docs for examples 👍

File: lib/fathead/nose2/parse.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env ruby
+require 'nokogiri'
+require 'open-uri'
+
+NOSE_HOME = "http://nose2.readthedocs.io/en/latest/getting_started.html"
+topics = ARGV
+
+topics.each do |topic|
+  url = "http://nose2.readthedocs.io/en/latest/getting_started.html##{topic}"
```
Comment (moollaza) [inline lib/fathead/nose2/parse.rb]: @rnkaufman this part of the documentation isn't ideal for a Fathead. Rather, I suggest you focus on http://nose2.readthedocs.io/en/latest/dev/event_reference.html and similar pages that define specific classes/functions/methods and other elements of the Nose2 API.

Those are each unique and specific, and when someone searches those we'll be able to show relevant documentation.

I rewrote a bunch of the Fathead documentation yesterday hoping to make this more clear.

Right now your Fathead only contains 4 articles, and the titles are very generic/vague/non-specific.

The article titles are the queries we expect to answer, so currently your fathead is going to show a result about Nose2 documentation when someone searches "dependencies" or "installation" -- clearly those results wouldn't be relevant to those queries.

You could still capture this content and prepend "nose2" to them so you'd have an article for "nose2 installation" -- that is way more specific!

**Comment by moollaza** [comment]
@rnkaufman Ping! Would you be able to finish this one up? We'd love to get this live!

**Comment by moollaza** [comment]
@rnkaufman I'm closing this due to inactivity. We can re-open it later if you're able to finish it up 👍

**Pull request closed without merging**